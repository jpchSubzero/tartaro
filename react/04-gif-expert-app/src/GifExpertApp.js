import React, {useState} from 'react';
import { AddCategory } from './components/AddCategory';
import { GifGrid } from './components/GifGrid';


// const GifExpertApp = () => {
    
const GifExpertApp = () => {
    //Esto no sirve porque no permite cambiar porque es una constante
    // const categories = ['One punch', 'Samurai X', 'Dragon Ball'];

    //Snippet -> useState
    // const [categories, setCategories] = useState(['One punch', 'Samurai X', 'Dragon Ball']);

    // Dejar solo una categoría para empezar a obtener los gifs desde el api
    const [categories, setCategories] = useState(['Saint Seiya']);


    // Tarea 1
        // Agregar un botón para aumentar categorías
        // Crear una función para procesar el evento click
        // Agregar un elemento estático al arreglo
        // Se debe presentar inmediatamente la nueva categoría

    // Quitamos porque se va a utilizar AddCategory
    // const handleAdd = () => {
    //     // No funciona pero no presenta ni warnings ni errores, el clg si presenta los cambios pero no se renderizan a pesar de ser un const ya que react no sabe que hubieron cambios
    //     // categories.push('Saint Seiya');
    //     // console.log(categories);

    //     // Esto da un error porque categories es un array y de esta forma lo convertimos en un string con el texto Saint Seiya, JS no se hace problema por cambios de tipos de datos pero al intentar un map() de un string salta el error
    //     // setCategories('Saint Seiya');

    //     // Forma larga: sacando una copia y agregándole el elemento dado que categorías es inmutable
    //     // let categoriesCopy = [...categories];
    //     // categoriesCopy.push('Saint Seiya');
    //     // setCategories(categoriesCopy);

    //     //Forma corta: destructurando categories y agregando el elemento, utilizar el setState forza la renderización
    //     // setCategories([...categories, 'Saint Seiya']);

    //     //En caso de querer una pila. Si se deja con el setCategories anterior solo toma en cuenta el último porque setCategories actualiza con la rederización
    //     // setCategories(['Saint Seiya', ...categories]);

    //     // También se puede usar un callback, función que puede recibir útil más adelante
    //     //En todos los casos solo funciona correctamente con el primer click porque luego da error porque se repite el key ya que es el mismo nombre de la categoría
    //     setCategories(categs => [...categs, 'Saint Seiya']);
    // }

    return (
        <>
            <h2>GifExpertApp</h2>

            {/* Para agregar el componente y visualizarlo, luego servirá para la tarea 3 */}
            {/* <AddCategory /> */}

            {/* Para poder agregar los elementos desde el componente hijo se envía como prop la función que se quiere ejecutar con el evento */}
            <AddCategory setCategories={ setCategories }/>

            <hr />

            {/* Reemplazar por el nuevo AddCategory */}
            {/* <button onClick={ handleAdd }>Agregar categoría</button> */}

            {/* Presenta todo junto pero por consola se ve que están separados */}
            {/* { categories } */}

            <ol>
                {/* No funciona porque entre {} se espera una expresión, un valor; el for es un ciclo y no es aceptado
                { for (category in categories) {<li>category</li>}} */}

                {/* El map devuelve un valor final renderizable directamente pero presenta un error por falta de key o identificador único*/}
                {/* { categories.map(category => <li>{category}</li>) } */}

                {/* Para solventar el problema se agrega el index como key pero no es recomendable porque si se agregan o eliminan elementos los keys se verán afectados */}
                {/* { categories.map((category, idx) => <li key={idx}>{category}</li>) } */}

                {/* Se puede utilizar el propio item como key */}
                {/* Dado que el key debe ser único no se podrían repetir elementos, en la mayoría de casos esta información viene de la base de datos y lo que se presentaría sería el id del elemento y su valor. Ej. { categories.map(category => <li key={category.id}>{category.name}</li>) }, así nunca se repetirían */}
                {/* { categories.map(category => <li key={category}>{category}</li>) } */}

                {/* Reemplazamos el li con el nuevo componente GifGrid */}
                { categories.map(category => 
                    <GifGrid 
                        key={ category }
                        category={ category } 
                    />
                ) }
            </ol>
        </>
    )
}

export default GifExpertApp;
