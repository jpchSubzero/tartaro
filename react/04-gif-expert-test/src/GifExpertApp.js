import React, {useState} from 'react';
import { AddCategory } from './components/AddCategory';
import { GifGrid } from './components/GifGrid';

// En este componente se puede validar que existan los elementos GifExpertApp y GifGrid.
// No se puede asignar un valor al state ya que no tiene un nombre y se puede inicializar varios states. Para demostrar, copiar varias veces 
// const [categories, setCategories] = useState(['Saint Seiya']); cambiando el nombre de categories y setCategories. En la consola de react se verá que existen varios useState con valores pero no se puede identificar cual quisieramos utilizar
// Para poder probar validando el state inicial se puede agregar un valor por defecto al componente como parámetro de entrada en lugar de ser un valor fijo (quemado), así no validamos directamente el state sino sus efectos

// const GifExpertApp = () => {
const GifExpertApp = ({ defaultCategories = [] }) => {

    // Implementar categorías por defecto, esto causará el error de snapshot
    const [categories, setCategories] = useState(defaultCategories);

    // const [categories, setCategories] = useState(['Saint Seiya']);

    // const [categories1, setCategories1] = useState(['Saint Seiya']);
    // const [categories2, setCategories2] = useState(['Saint Seiya']);
    // const [categories3, setCategories3] = useState(['Saint Seiya']);

    return (
        <>
            <h2>GifExpertApp</h2>

            <AddCategory setCategories={ setCategories }/>

            <hr />

            <ol>
                { categories.map(category => 
                    <GifGrid 
                        key={ category }
                        category={ category } 
                    />
                ) }
            </ol>
        </>
    )
}

export default GifExpertApp;

// Tarea 1
// Validar a través de snapshot