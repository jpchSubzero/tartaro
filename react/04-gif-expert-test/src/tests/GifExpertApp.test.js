import GifExpertApp from "../GifExpertApp";
import { shallow } from 'enzyme';

describe('Pruebas con el componente GifExpertApp', () => {
    test('Debe ser igual que el snapshot', () => {
        const wrapper = shallow(<GifExpertApp />);
        expect(wrapper).toMatchSnapshot();
    });
    
    // Una vez implementada la categoría por defecto podemos validarla
    test('Debe mostrar una lista de categorías', () => {
        const categories = ['saint seiya', 'one punch', 'dragon ball z'];

        const wrapper = shallow(<GifExpertApp defaultCategories = { categories }/>);
        
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('GifGrid').length).toBe(categories.length);
    });
   
})
