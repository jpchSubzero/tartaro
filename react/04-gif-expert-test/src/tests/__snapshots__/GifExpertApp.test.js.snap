// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas con el componente GifExpertApp Debe mostrar una lista de categor√≠as 1`] = `
<Fragment>
  <h2>
    GifExpertApp
  </h2>
  <AddCategory
    setCategories={[Function]}
  />
  <hr />
  <ol>
    <GifGrid
      category="saint seiya"
      key="saint seiya"
    />
    <GifGrid
      category="one punch"
      key="one punch"
    />
    <GifGrid
      category="dragon ball z"
      key="dragon ball z"
    />
  </ol>
</Fragment>
`;

exports[`Pruebas con el componente GifExpertApp Debe ser igual que el snapshot 1`] = `
<Fragment>
  <h2>
    GifExpertApp
  </h2>
  <AddCategory
    setCategories={[Function]}
  />
  <hr />
  <ol />
</Fragment>
`;
